### PR TITLE
ceph: add a procedure to recover mon from OSDs

### PR DESCRIPTION
**Pull request by llamerada-jp**
Signed-off-by: Yuji Ito <llamerada.jp@gmail.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: Documentation/ceph-disaster-recovery.md [hunk 1]
```diff
@@ -229,6 +229,170 @@ kubectl -n rook-ceph scale deployment rook-ceph-operator --replicas=1
...
+```console
+TOOL=$(kubectl get -n ${NAMESPACE} po -l app=rook-ceph-tools -o json | jq -r '.items[0].metadata.name')
+
+kubectl exec -i -n ${NAMESPACE} ${TOOL} -- ceph mon enable-msgr2
+
+kubectl -n ${NAMESPACE} scale deployment rook-ceph-operator --replicas=1
+
+kubectl -n ${NAMESPACE} delete pod -l app=rook-ceph-mgr
+
+# Check if cluster is recovered
```
Comment (satoru-takeuchi) [inline Documentation/ceph-disaster-recovery.md:386]: Check if cluster becomes healthy.

File: Documentation/ceph-disaster-recovery.md [hunk 2]
```diff
@@ -229,6 +229,170 @@ kubectl -n rook-ceph scale deployment rook-ceph-operator --replicas=1
...
+kubectl exec -n ${NAMESPACE} ${TOOL} -- ceph -s
+```
+
+### After recover
+
+After following the above steps, the mon will be restored and you can read the data.
+However, the data that is not synced to OSD is lost. Also, some settings in mgr and ceph cluster are not restored.
+It is recommended that you take a copy from the recovered cluster and move the data to the newly recreated cluster.
```
Comment (satoru-takeuchi) [inline Documentation/ceph-disaster-recovery.md:394]: この節はあえて書かなくてもいいと思います。

**Review by satoru-takeuchi — CHANGES_REQUESTED** [review]

File: Documentation/ceph-disaster-recovery.md [hunk 3]
```diff
@@ -229,6 +229,170 @@ kubectl -n rook-ceph scale deployment rook-ceph-operator --replicas=1
 
 The operator will automatically add more mons to increase the quorum size again, depending on the `mon.count`.
 
+
+## Recovering Mon Quorum using OSDs
+
+In some very unfortunate cases, the data that all MONs have is damaged and they cannot be booted.
+In the following method, we will try to recover a MON based on the data stored by OSD.
+See: https://docs.ceph.com/en/latest/rados/troubleshooting/troubleshooting-mon/#recovery-using-osds
```
[the remark this answers is not in the document]
Comment (toshipp) [reply Documentation/ceph-disaster-recovery.md:237]: この辺にリンク先の known issue があるので気を付けてねというのも書いておきたいですね。
こんな感じかな。

It is also that known issues described in the ceph document exist, please read it before and carry out the work carefully.

File: Documentation/ceph-disaster-recovery.md [hunk 4]
```diff
@@ -229,6 +229,170 @@ kubectl -n rook-ceph scale deployment rook-ceph-operator --replicas=1
...
+done
+
+for osd in ${OSDS}; do
+    kubectl -n ${NAMESPACE} get deployment $osd -o yaml > ${BACKUP_DIR}/$osd.yaml
+done
+
+kubectl -n ${NAMESPACE} get deployment ${TOOL_DEPLOYMENT} -o yaml > ${BACKUP_DIR}/${TOOL_DEPLOYMENT}.yaml
+```
+
+Stop the operator, mons and OSDs:
```
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply Documentation/ceph-disaster-recovery.md:275]: 自分も MON じゃないかなと考えましたが、同ページの他の場所で mon が使われていたのでそちらにあわせてました。

File: Documentation/ceph-disaster-recovery.md (hunk 2 above)
Comment (llamerada-jp) [reply Documentation/ceph-disaster-recovery.md]: MGRの設定が復元されない点は残したほうが良いかと思い、240行目に移動しました。
https://github.com/cybozu-go/rook/pull/4/files#diff-0ca833866b2d53e3e162d867120a0b54da02afef2b3b10b276cbda17bf95302eR240

**Review by toshipp — APPROVED** [review]
LGTM

**Pull request closed without merging**